# 02 - Python basics B

- 字符串 strings
- 流程控制 control flow
- 可迭代对象、迭代器 iterable & iterator
- 表达式 comprehensions
- 文件对象 file object

## 字符串Strings

### 创建

In [1]:
# single quote/double是一样的
"string" == 'string'

True

In [4]:
"string'", 'string"'

("string'", 'string"')

In [7]:
"""string'""", '''string" '''

("string'", 'string" ')

In [8]:
"""
abc
def
"""

'\nabc\ndef\n'

In [9]:
len("next")

4

In [20]:
len("\next")

4

**转义字符(escape character)**: 反斜杠 `\` 开头的特殊字符


| 转义字符 | 说明                          | 示例                | 输出结果          |
|----------|-----------------------------|---------------------|------------------|
| `\\`     | 反斜杠                      | `"C:\\Users"`       | `C:\Users`       |
| `\'`     | 单引号                      | `'It\'s me'`       | `It's me`        |
| `\"`     | 双引号                      | `"She said \"Hi\""` | `She said "Hi"`  |
| `\n`     | 换行符                      | `"Hello\nWorld"`   | `Hello`<br>`World` |
| `\t`     | 水平制表符 (Tab)            | `"Name:\tAlice"`   | `Name:    Alice` |


In [14]:
"She said \"Hi\""

'She said "Hi"'

In [15]:
print("She said \"Hi\"")

She said "Hi"


In [16]:
a = "C:\\data"
a  # 以代码的形式输出

'C:\\data'

In [17]:
print(a)

C:\data


In [18]:
# 三重引号创建的字符串支持换行
x = """
This is a string.
"""
x

'\nThis is a string.\n'

**原始字符串(raw-string)** 是以字母`r`开头的特殊字符串，转义机制在这里不起作用。

In [19]:
print(r'She said \"Hi\"')

She said \"Hi\"


In [21]:
len(r'\next')

5

### 基本操作

In [22]:
'abc' + 'def'

'abcdef'

In [26]:
'abc' 'def' 'ghi'

'abcdefghi'

In [27]:
name = "LI Lei"
print("Hello, " + name + "!")

Hello, LI Lei!


In [28]:
print(f"Hello, {name}!")

Hello, LI Lei!


In [32]:
'abc' * 3

'abcabcabc'

In [33]:
list('abc')

['a', 'b', 'c']

In [34]:
list('abc') * 2

['a', 'b', 'c', 'a', 'b', 'c']

In [35]:
'abc'[::-1]

'cba'

In [36]:
'a' < 'b', 'a' == 'a', 'a ' > 'a'

(True, True, True)

补充：方法（method）即绑定某个对象的函数
- `object.attribute`
- `function(arguments)`

`object.method(arguments)`

In [37]:
# 列出x对象的所有attribute
print(dir(x))

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'removeprefix', 'removesuffix', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


In [56]:
x = 'abc-abc-abd-a-b'
x.replace('ab', 'xy')

'xyc-xyc-xyd-a-b'

In [41]:
x.replace('ab', 'xy', 1)

'xyc-abc-abd'

In [42]:
x.capitalize()

'Abc-abc-abd'

In [44]:
x.find('c')

2

In [45]:
'123'.isdecimal()

True

In [46]:
'123'.isdigit()

True

In [48]:
'    \t\n'.isspace()

True

In [49]:
','.join(['123', '456', '789'])

'123,456,789'

In [50]:
'123,456,789'.split(',')

['123', '456', '789']

In [51]:
'123,456,789'.split(',', 1)

['123', '456,789']

In [52]:
'123,456,789'.rsplit(',', 1)  # r = right

['123,456', '789']

In [53]:
'abc'.startswith('ab'), 'abc'.endswith('x')

(True, False)

In [54]:
'abc'.upper(), 'Abc'.lower()

('ABC', 'abc')

In [57]:
'aabbccabc'.translate(str.maketrans('abc', 'xyz'))

'xxyyzzxyz'

In [65]:
x = 'abc\t xyz\t3.14\trst\n'
print(x)

abc	 xyz	3.14	rst



In [66]:
x.split()

['abc', 'xyz', '3.14', 'rst']

#### 字符串格式化

(已经讲过f-string)

In [60]:
"Hello, {} and {}!".format(name, "HAN Meimei")

'Hello, LI Lei and HAN Meimei!'

In [61]:
"pi = %f, r = %f" % (3.1, 2.1324)

'pi = 3.100000, r = 2.132400'

In [62]:
"pi = %05.2f, r = %06.02f" % (3.1, 2.1324)

'pi = 03.10, r = 002.13'

In [63]:
# "02"表示小数点后面精确到2位小数，不足在后面补0
# "05"中的5表示总的字符串宽度，0表示不足在前面补0
len("%05.2f" % (3.1))

5

#### CSV与TSV的文本解析

In [68]:
x = 'abc,xyz,3.14,rst  \n'
x.split(',')

['abc', 'xyz', '3.14', 'rst  \n']

In [69]:
x.rstrip()

'abc,xyz,3.14,rst'

In [70]:
x.rstrip().split(',')

['abc', 'xyz', '3.14', 'rst']

### 正则表达式 (regular expression)

python对正则表达式的支持：
- https://docs.python.org/3/library/re.html
- https://docs.python.org/3/howto/regex.html#regex-howto

怎么写正则表达式：LLM可能比大多数程序员效率高。

🙋**练习：自学python正则表达式**
- https://realpython.com/regex-python
- https://realpython.com/regex-python-part-2/
- https://www.runoob.com/python3/python3-reg-expressions.html （中文）

## 函数(function)也是对象

```python
def function_name(arg1, arg2, ...):
    ... 
    return 
```

In [ ]:
def function_name(arg1, arg2):
    """
    documentation
    """
    result = [arg1, arg2]
    return result

## 对象的特性(attributes)

```python
object.attribute
```
python中的对象有很多特性，特性可以是普通的对象，即属性(property)，也可以是适用于该对象的函数,即方法（method）。

```python
object.__attribute__
```
前后都有两个下划线的特性，是特殊的特性，一般由python代码解释器调用，而非用户直接调用。

例如这些方法在冒号后面对应的指令出现时，由代码解释器隐式调用：
- `x.__iter__`: `iter(x)`
- `x.__str__`: `print(x)`
- `x.__len__`: `len(x)`
- `x.__contains__`: `i in x`
- `x.__getitem__(slice(start, stop, step))`: `x[start:stop:step]`

In [71]:
# dir函数可以查看函数的所有特性
help(dir)

Help on built-in function dir in module builtins:

dir(...)
    dir([object]) -> list of strings

    If called without an argument, return the names in the current scope.
    Else, return an alphabetized list of names comprising (some of) the attributes
    of the given object, and of attributes reachable from it.
    If the object supplies a method named __dir__, it will be used; otherwise
    the default dir() logic is used and returns:
      for a module object: the module's attributes.
      for a class object:  its attributes, and recursively the attributes
        of its bases.
      for any other object: its attributes, its class's attributes, and
        recursively the attributes of its class's base classes.



In [72]:
# print函数会通过对象的.__str__获取对象的string表示形式
# 如果没有定义.__str__会使用.__repr__
# 如果都没定义，会显示类型跟内存地址
x = ['a', 'b', 3]
print(x)

['a', 'b', 3]


In [73]:
x.__str__()

"['a', 'b', 3]"

In [74]:
str(x)

"['a', 'b', 3]"

## 常用的python指令 (statement)


##### 基础语句

| 指令 | 说明 | 示例 |
|------|------|------|
| `=` | 赋值语句 | `x = 10` |
| `del` | 删除变量或对象元素 | `del lst[0]` |
| `pass` | 空操作占位符 | `def foo(): pass` |
| `assert` | 断言条件为真 | `assert x > 0` |
| `raise` | 抛出异常 | `raise ValueError()` |

##### 流程控制

| 指令 | 说明 | 示例 |
|------|------|------|
| `if/elif/else` | 条件判断 | `if x>0: print("正数")` |
| `for` | 迭代循环 | `for i in range(5):` |
| `while` | 条件循环 | `while x < 10:` |
| `break` | 跳出循环 | `while True: break` |
| `continue` | 跳过当前迭代 | `for i in range(5): if i==2: continue` |
| `try/except/finally` | 异常处理 | `try: x=1/0 except: print("错误")` |

##### 函数与类

| 指令 | 说明 | 示例 |
|------|------|------|
| `def` | 定义函数 | `def foo(): return 1` |
| `return` | 函数返回值 | `return x + y` |
| `yield` | 生成器返回值 | `yield i**2` |
| `class` | 定义类 | `class Foo:` |
| `lambda` | 匿名函数 | `f = lambda x: x+1` |

##### 导入与作用域

| 指令 | 说明 | 示例 |
|------|------|------|
| `import` | 导入模块 | `import os` |
| `from...import` | 从模块导入 | `from math import sqrt` |
| `global` | 声明全局变量 | `global count` |
| `nonlocal` | 声明非局部变量 | `nonlocal x` |

##### 上下文管理

| 指令 | 说明 | 示例 |
|------|------|------|
| `with` | 上下文管理器 | `with open('a.txt') as f:` |
| `as` | 别名赋值 | `import numpy as np` |

##### 其他特殊语句

| 指令 | 说明 | 示例 |
|------|------|------|
| `match/case` (Python 3.10+) | 模式匹配 | `match x: case 1: print("one")` |
| `...` (Ellipsis) | 省略号对象 | `def foo(): ...` |
| `@` | 装饰器语法 | `@decorator def foo(): pass` |


### 赋值语句

- 赋值语句将对象与变量名绑定；
- 变量名存储的不是对象的值，而是对象的内存地址，即变量是对 对象 的引用(reference)，类似于C语言中的指针；

In [75]:
# basic assignment
a = [1, 2, 3]
a

[1, 2, 3]

In [76]:
# sequence assignment
a, b, c = [1, 2, 3]
print(f'a = {a}, b = {b}, c = {c}')

a = 1, b = 2, c = 3


In [77]:
[a, b, c] = [1, 2, 3]
print(f'a = {a}, b = {b}, c = {c}')

a = 1, b = 2, c = 3


In [78]:
a, b, c = range(3)
print(f'a = {a}, b = {b}, c = {c}')

a = 0, b = 1, c = 2


In [80]:
list(range(3))

[0, 1, 2]

In [81]:
a, b, c = range(2)
print(f'a = {a}, b = {b}, c = {c}')

ValueError: not enough values to unpack (expected 3, got 2)

In [83]:
a, _ = ('2025', 'zhang san')  # _ 常用来匹配某个位置，但其值后续不再用到
print(a)

2025


In [85]:
# extended-unpacking assignment
a, b = [1, 2, 3]

ValueError: too many values to unpack (expected 2)

In [86]:
a, *b = [1, 2, 3]
a, b

(1, [2, 3])

In [87]:
a, *b, c = [1, 2, 3, 4]
a, b, c

(1, [2, 3], 4)

In [88]:
a, b, *c = range(2)
a, b, c

(0, 1, [])

In [89]:
a, b, *c, *d = range(2)
a, b, c

SyntaxError: multiple starred expressions in assignment (991317121.py, line 1)

In [90]:
a = b = c = d = 5
a += 1   # a = a + 1
b *= 2   # b = b * 2
c /= 2   # c = c / 3
d %= 3   # d = d % 3
a, b, c, d

(6, 10, 2.5, 2)

In [92]:
# assignment in expression
b = (a := 3) + 4  # equal to: a = 3; b = a + 3
a, b

(3, 7)

##### 🙋练习

1. 在执行之前你能猜出下面这两段代码的输出是什么吗？
```python
a, *b = 'python'
print([a, b])
```

```python
a = []
a += 'python'
print(a)
```

2.  已知=右边是`(110, (119, 120))`，怎么用一行赋值语句把三个数值分别赋给a, b, c?
```python
? = (110, (119, 120))
```

##### 共享的引用

In [93]:
a = [1, 2, 3]
b = a
print(b)
a[0] = -1
print(b)

[1, 2, 3]
[-1, 2, 3]


In [94]:
a = [1, 2, 3]
b = [a, 4, 5, 6]
print(b)

[[1, 2, 3], 4, 5, 6]


In [95]:
a[0] = -1
print(b)

[[-1, 2, 3], 4, 5, 6]


In [98]:
# 怎么避免
a = [1, 2, 3]
b = [a[:], 4, 5, 6]
print(b)
a[0] = -1
print(b)
print(a)

[[1, 2, 3], 4, 5, 6]
[[1, 2, 3], 4, 5, 6]
[-1, 2, 3]


In [99]:
a = [1, 2, 3]
b = a.copy()
print(b)
a[0] = -1
print(b)

[1, 2, 3]
[1, 2, 3]


### print

In [100]:
print('a', 'b')

a b


In [102]:
print('a', 'b', sep=',')  # sep is short for seperator

a,b


In [103]:
print('a', 'b', 'c', sep='\t')
print('-----')

a	b	c
-----


In [106]:
print('a', 'b', 'c', sep='\t', end=';')
print('-----')

a	b	c;-----


### 流程控制

开始之前：
- python会自动检测指令或代码块结束的位置；
- 一行的末尾表示一条指令的结束，不需要用特殊的符号比如`;`来结束一条指令；
- 可以用`;`把几个简单指令组合写在一行；
- 如果指令延续多行，可以用括号(`()`, `[]`, `{}`)括起来；
- 在一行的末尾加一个反斜杠也可以表示指令还没结束，需要延续到下一行，但容易出错；(不推荐使用)
- 缩进表示一个代码块(code block)的开始，缩进结束即block的结束；（注意: 在正式的文本中不要中英混杂）；
- 缩进的形式可以用tab、四个空格、两个空格、或tab，但在程序中要保持一致，不要混合使用；

In [107]:
x = [1, 2,
     3, 4]
x

[1, 2, 3, 4]

In [112]:
y = 'a' 'b'
'c'
print(y)

ab


In [113]:
y = ('a' 'b'
     'c')
print(y)

abc


#### if-else

```python
if test:
    code block
elif test:
    code block
else:
    code block
```

In [114]:
x = 3
if x > 0:
    print("x is positive!")

x is positive!


In [115]:
x = -2
if x > 0:
    print("x is positive!")
else:
    print("x is not positive!")

x is not positive!


In [116]:
x > 0

False

In [117]:
x = -2
if x > 0:
    print("x is positive!")
elif x == 0:
    print("x is zero!")
else:
    print("x is negative!")

x is negative!


In [118]:
x = 0
if x > 0:
    print("x is positive!")
elif x == 0:
    print("x is zero!")
else:
    print("x is negative!")

x is zero!


In [121]:
# 类似于C语言中的条件赋值如何实现？
#  C : condition? a : b
#  py: a condition? else b
a = 2
b = 'positive' if a > 0 else 'non-positive'
print(b)

positive


In [122]:
# multiple choices
a = 2
match a:
    case 1:
        print('first option')
    case 2:
        print('second option')
    case 3:
        print('third option')

second option


In [123]:
a = 7
match a:
    case 1:
        print('first option')
    case 2:
        print('second option')
    case 3:
        print('third option')
    case _:
        print('invalid option')


invalid option


🙋练习

尝试理解这段代码，并预测其输出。
```python
scores = [('zhang san', 61), ('li si', 85), ('wolong', 48), ('zhuge', 100)]
for record in scores:
    match record:
        case (name, score) if score >= 60:
            print(f'    {name}: pass')
        case (name, _):
            print(f'    {name}, fail')
```

In [124]:
score_a, score_b, score_c = 100, 80, 61
if score_a >= 60 and score_b >= 60 and score_c >= 60:
    print('pass')
else:
    print('fail')

pass


In [125]:
# 大小比较返回True，False (即返回整数1或0）
2 < 3, 2 > 3

(True, False)

In [127]:
# `and` 返回什么？
a = score_a > 60 and score_b > 60
a

True

In [128]:
a = 2 and 3
print(a)

3


In [129]:
a = 3 and 2
print(a)

2


##### 反思and、or、not

- python中的任何对象都有一个内在的`True`或`False`值；
- 数字0、空对象、`None`的布尔值是`False`，其他对象的布尔值是`True`；
- 比较(>, <, >=, <=)或相等（==）测试返回的值是`True`/`False`；
- `or` : 如果左侧对象的布尔值是`True`，返回左边的对象，否则返回右边的对象；
- `and`: 如果左侧对象的布尔值是`False`，返回左边的对象，否则返回右边的对象；
- `not`: 返回与右侧对象的布尔值相反的布尔值；

In [136]:
x = 2 or 3
print(x)

2


In [137]:
x = 0 or 2
x

2

🙋**练习**

1. 判断下面代码的数据结果
```python
a = 0 or 1
print(a)
```

2. 哪个对象特性决定了对象的内在布尔值是True还是False？

##### is和==

- `==` 比较两个对象的值是否相等；
- `is` 比较两个对象是否相同（位于相同的内存地址）；

In [138]:
a = b = 'Hello, World!'
c = 'Hello, World!'
a is b, a is c

(True, False)

In [139]:
# 查看a,b,c的内存地址
id(a), id(b), id(c)

(4430168560, 4430168560, 4436364912)

In [140]:
a == b, a == c

(True, True)

🙋**练习**

判断下面代码的数据结果是True还是False？
```python
a = 'py'
b = 'py'
print(a is b)
```
一定要做一下，没有看起来那么简单！想想为什么会输出这样的结果？(提示：用`id`检验自己的猜想)

#### While

```python
while test:
    ...
    if test: break
    if test: continue
else:
    ...  # 循环正常结束时(即不通过break结束）执行
```

注意：**不要写死循环！**

In [ ]:
# do not run this cell in notebook
while True:
    x = input("Please enter a number:")
    if x == "q": break
    print(float(x)**2)
print("Bye!")

In [141]:
y = []
x = 1
while x < 10:
    print(f'x = {x}', end=', ')
    y.append(x**2)
    x += 1

y

x = 1, x = 2, x = 3, x = 4, x = 5, x = 6, x = 7, x = 8, x = 9, 

[1, 4, 9, 16, 25, 36, 49, 64, 81]

In [142]:
from time import sleep

y = []
x = 1

while x < 10:
    print(f'x = {x}', end=', ')
    y.append(x**2)
    # x += 1
    sleep(1)

y

x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, x = 1, 

KeyboardInterrupt: 

#### for-loop

```python
for item in iterable:
    ...
    if test: break
    if test: continue
    ...
else:
    ...
```

可迭代对象(iterable)是可以遍历的对象，本质上来说是实现了`__iter__`方法的对象，包括各种序列(list, tuple, range)以及dict等以及迭代器。

In [143]:
x = [1, 'a', 'abc']
'__iter__' in dir(x)

True

In [144]:
y = range(10)
'__iter__' in dir(y)

True

In [145]:
for i in [1, 'a', 'abc']:
    print(i)

1
a
abc


In [150]:
x = dict(zip('abc', ['1', '2', '3']))
x

{'a': '1', 'b': '2', 'c': '3'}

In [151]:
for i in x:
    print(i)

a
b
c


In [152]:
x.keys()

dict_keys(['a', 'b', 'c'])

In [153]:
x.items()

dict_items([('a', '1'), ('b', '2'), ('c', '3')])

In [149]:
x = []
for i in range(10):
    print(f'i = {i}')
    if i % 2 == 1:
        continue   # 如果是奇数，当前循环结束，继续下一个循环
    x.append(i**2)

x

i = 0
i = 1
i = 2
i = 3
i = 4
i = 5
i = 6
i = 7
i = 8
i = 9


[0, 4, 16, 36, 64]

In [154]:
d = {'a': 1, 2: 'b'}
for key, value in d.items():
    print(f'{key} => {value}')

a => 1
2 => b


##### for-loop相关的辅助函数

```
range(from, to, step_size=1)

zip(iterable, iterable, ...)

enumerate(iterable)
```

In [159]:
range(10)

range(0, 10)

In [160]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [161]:
list(range(1, 10))

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [162]:
list(range(1, 10, 2))

[1, 3, 5, 7, 9]

In [163]:
x = range(5)
y = 'abcde'
zip(x, y)

In [164]:
list(zip(x, y))

[(0, 'a'), (1, 'b'), (2, 'c'), (3, 'd'), (4, 'e')]

In [165]:
# length is determined by the shortest iterable
list(zip(range(3), 'abcde'))

[(0, 'a'), (1, 'b'), (2, 'c')]

In [166]:
z = list('ABCDE')

for i in zip(x, y, z):
    print(i)

(0, 'a', 'A')
(1, 'b', 'B')
(2, 'c', 'C')
(3, 'd', 'D')
(4, 'e', 'E')


In [167]:
for a, *b in zip(x, y, z):
    print(a, b)

0 ['a', 'A']
1 ['b', 'B']
2 ['c', 'C']
3 ['d', 'D']
4 ['e', 'E']


In [168]:
index = 0
for char in 'abcde':
    print(f'character at index {index} is {char}')
    index += 1

character at index 0 is a
character at index 1 is b
character at index 2 is c
character at index 3 is d
character at index 4 is e


In [169]:
for i in enumerate('abcde'):
    print(i)

(0, 'a')
(1, 'b')
(2, 'c')
(3, 'd')
(4, 'e')


In [170]:
for index, char in enumerate('abcde'):
    print(f'character at index {index} is {char}')

character at index 0 is a
character at index 1 is b
character at index 2 is c
character at index 3 is d
character at index 4 is e


#### try-except
捕获可能出现的代码错误，并采取必要的措施，防止程序中断；

In [171]:
print(1/0)

ZeroDivisionError: division by zero

In [172]:
try:
    print(1/0)
except:
    print('Division error')

Division error


In [176]:
x = 1
y = 0
try:
    z = x/y
except ZeroDivisionError:
    print('division by zero')
    z = 100

division by zero


In [177]:
print(z)

100


In [178]:
d = dict()
try:
    x = d['A']
except KeyError:
    print('"A" is not a valid key in d')

"A" is not a valid key in d


## 可迭代对象（iterable）

- 可迭代对象（iterable）: 有`__iter__`方法的对象，可以遍历生成序列；
- 迭代器（iterator）: 有`__next__`方法的对象，每次运行都会给出对象下一次迭代的值；


In [179]:
x = enumerate(range(1, 5))
', '.join(dir(x))

'__class__, __class_getitem__, __delattr__, __dir__, __doc__, __eq__, __format__, __ge__, __getattribute__, __getstate__, __gt__, __hash__, __init__, __init_subclass__, __iter__, __le__, __lt__, __ne__, __new__, __next__, __reduce__, __reduce_ex__, __repr__, __setattr__, __sizeof__, __str__, __subclasshook__'

In [180]:
list(x)

[(0, 1), (1, 2), (2, 3), (3, 4)]

In [181]:
x = enumerate(range(1, 5))
a = list(x)
b = list(x)
a == b

False

In [182]:
a

[(0, 1), (1, 2), (2, 3), (3, 4)]

In [183]:
b

[]

In [184]:
next(x)

StopIteration: 

In [185]:
x = enumerate(range(1, 5))

In [186]:
next(x)

(0, 1)

In [187]:
next(x)

(1, 2)

In [188]:
list(x)  # 迭代器中已经遍历过的值没了！

[(2, 3), (3, 4)]

In [ ]:
list(x)  # # 迭代器会耗竭

In [189]:
list(x)

[]

In [190]:
x = 'abcde'
y = iter(x)  # 把iterable转为iterator

In [191]:
next(x)

TypeError: 'str' object is not an iterator

In [192]:
next(y)

'a'

In [193]:
next(y)

'b'

In [194]:
list(x), list(y)

(['a', 'b', 'c', 'd', 'e'], ['c', 'd', 'e'])

In [195]:
x = 'aaabbbcccdddeeefff'
list(map(''.join, zip(*[iter(x)]*3)))
# 练习题：分解代码，逐步执行，看看怎样一步步得到这个结果的

['aaa', 'bbb', 'ccc', 'ddd', 'eee', 'fff']

##### 函数迭代(functional itrable)

In [196]:
x = map(int, '12345')

In [197]:
x

In [198]:
', '.join(dir(x))

'__class__, __delattr__, __dir__, __doc__, __eq__, __format__, __ge__, __getattribute__, __getstate__, __gt__, __hash__, __init__, __init_subclass__, __iter__, __le__, __lt__, __ne__, __new__, __next__, __reduce__, __reduce_ex__, __repr__, __setattr__, __sizeof__, __str__, __subclasshook__'

In [200]:
print(next(x))

2


In [201]:
list(x)

[3, 4, 5]

## Python 列表推导式（List Comprehensions）

In [ ]:
y = []
for i in range(10):
    y.append(i**2)
y

In [ ]:
[i**2 for i in range(10)]

In [ ]:
%%timeit -n 10 -r 5
# -n: 每轮重复执行多少次代码；
# -r: 一共执行几轮；
y = []
for i in range(100_0000):
    y.append(i**2)

In [ ]:
%timeit -n 10 -r 5 [i**2 for i in range(100_0000)]

In [ ]:
y = []
for i in range(10):
    if i % 2 == 0:
        y.append(i**2)
y

In [ ]:
# with if
[i**2 for i in range(10) if i % 2 == 0]

In [ ]:
# nested for loop
x = [i + j + k for i in 'ACGT' for j in 'ACGT' for k in 'ACGT']
print(x)

In [ ]:
print([i for i in x if not i in ['TAA', 'TAG', 'TGA']])

In [ ]:
# 结果可以是其他类型，不限于list
{i**2 for i in range(10)}

In [ ]:
{i:i**2 for i in range(10)}

In [ ]:
(i**2 for i in range(10))  # generator, not tuple

### Generator (生成器）与 Iterator（迭代器）的区别
生成器（Generator）是一种特殊的迭代器，它允许你根据需要一个一个生成值（惰性计算），而不是一次性生成所有数据。这种特性使其特别适合处理大规模数据流或无限序列，能显著节省内存。

In [ ]:
import sys

lst = [x for x in range(100000)]
print(sys.getsizeof(lst))

# 生成器表达式 （generator expression)
gen = (x for x in range(100000))
print(sys.getsizeof(gen))

🙋**练习**

下面两段代码的输出一样吗？为什么？

代码1：
```python
x = [[0] * 2 for i in range(2)]
x[0][0] = 1
print(x)
```

代码2：
```python
x = [[0] * 2] * 2
x[0][0] = 1
print(x)
```

## 扩展 （自学内容）

### 基础对象的拓展

- https://docs.python.org/3/library/collections.html


###  为高效循环创建迭代器的函数

- https://docs.python.org/3/library/itertools.html (english)
- https://docs.python.org/zh-cn/3.12/library/itertools.html (中文）

## File

In [ ]:
# 创建一个名为tmp的临时文件夹，可以单击右侧的新增文件夹按钮手动创建
! mkdir -p tmp

In [ ]:
f = open('tmp/test.txt', 'w')

In [ ]:
f

In [ ]:
', '.join(dir(f))

In [ ]:
f.write('abc')

虽然我们命令python在文件中写入"abc",但是实际文件仍然是空的，因为python为了提高效率，将输出暂时缓存，超过一定大小时才会真正写入。

In [ ]:
# like将缓存的内容（buffer）写入到文件
f.flush()

In [ ]:
f.writelines(['1\n', '2\n', '3\n'])

In [ ]:
f.flush()

In [ ]:
# 关闭文件对象
f.close()

In [ ]:
f.write('a new line')

In [ ]:
f2 = open('tmp/test.txt', 'r')

In [ ]:
f2

open函数的第二个参数是处理文件的模式
- `r`: read，读取；
- `w`: write，写入；
- `rb` / `wb`: 以二进制字节（byte）而非字符形式读/写；
- `r+`: 即可读也可写；

In [ ]:
f2.read()

In [ ]:
f2.close()

In [ ]:
f2 = open('tmp/test.txt', 'r')
f2.readline()

In [ ]:
next(f2)

In [ ]:
next(f2)

In [ ]:
next(f2)

In [ ]:
f2.closed

In [ ]:
f2.close()

In [ ]:
f3 = open('tmp/test.txt', 'r')
for line in f3:
    print(line)

In [ ]:
f3.closed

In [ ]:
f3.close()

文件对象不关闭会占用系统资源，但是老是忘记关闭怎么办？

> 使用context manager ！

Context Manager（上下文管理器）是 Python 中一种自动管理资源的机制，用于确保资源（如文件、网络连接、数据库连接、锁等）在使用后正确释放，避免泄漏或错误。

In [ ]:
with open('tmp/test.txt', 'r') as f3:
    for line in f3:
        print(line, end='')

In [ ]:
f3.closed

## Homework

fasta是最常用的生物信息学文件格式之一，其具体形式如下：
```
>id
sequence
>id
sequence
```

读取`homework/seqs.fasta`,计算每条序列的长度，输出一个名为`seqs_length.tsv`的tsv文件，第一列为序列id，第二列为序列长度：
```
id    length
ad    14
...
```
注意：
- 不能使用任何第三方module/package;
- 生成式AI仅限于查询内置函数的具体用法、报错信息的意义，千万不要复制题目让AI直接给出完整解决方案；

## 拓展阅读材料

《Fluent Python， 2nd edition》第一部分关于Python数据结构的深入介绍；